soc: st: Add support for STOP mode on STM32H5 devices

ABESTM opened the pull request:
Adds support for STOP mode on STM32H5 devices.
Also adds support for STM32H5 to LPTIM driver.
The functionality can be tested in samples/boards/stm32/power_mgmt/blinky and samples/boards/stm32/power_mgmt/serial_wakeup example.

ABESTM committed:
soc: st: add PM support for STM32H5

Single STOP mode level supported on STM32H5

Signed-off-by: Adam Berlinger <adam.berlinger@st.com>

ABESTM committed:
drivers: timer: add support for LPTIM on STM32H5

Added support for LPTIM1/2 for STM32H503 and
LPTIM1 to LPTIM6 for STM32H56x/STM32H57x devices

Signed-off-by: Adam Berlinger <adam.berlinger@st.com>

ABESTM committed:
boards: st: add support for LPTIMER for nucleo_h563zi board

LPTIM4 is selected, since it is used only for internal timing
and it doesn't support output/input

Signed-off-by: Adam Berlinger <adam.berlinger@st.com>

erwango commented on soc/st/stm32/stm32h5x/power.c:
```diff
@@ -0,0 +1,79 @@
+#include <soc.h>
+#include <zephyr/init.h>
+
+#include <stm32h5xx_ll_utils.h>
+#include <stm32h5xx_ll_bus.h>
+#include <stm32h5xx_ll_cortex.h>
+#include <stm32h5xx_ll_pwr.h>
+#include <stm32h5xx_ll_system.h>
```
Are all these files actually required ?
At first sight, I would say that utils, bus and system could be removed ?

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -403,8 +403,10 @@ static int sys_clock_driver_init(void)
 #if defined(LL_APB1_GRP1_PERIPH_LPTIM1)
 	LL_APB1_GRP1_ReleaseReset(LL_APB1_GRP1_PERIPH_LPTIM1);
-#elif defined(LL_APB3_GRP1_PERIPH_LPTIM1)
+#elif defined(LL_SRDAMR_GRP1_PERIPH_LPTIM1AMEN)
 	LL_SRDAMR_GRP1_EnableAutonomousClock(LL_SRDAMR_GRP1_PERIPH_LPTIM1AMEN);
```
This specific change is actually a fix over the initial implementation which impacts U5. Can you make it a specific commit ?
Also, since both `LL_APB3_GRP1_PERIPH_LPTIM1` and `LL_SRDAMR_GRP1_PERIPH_LPTIM1AMEN` are defined in STM32U5, the if/elif/elif sequence would not be valid anymore on U5
I'd propose to split out `LL_SRDAMR_GRP1_EnableAutonomousClock(LL_SRDAMR_GRP1_PERIPH_LPTIM1AMEN);` to a dedicated `#if defined(LL_SRDAMR_GRP1_PERIPH_LPTIM1AMEN)` section, then validate that running following sequence is fine on U5:
```
LL_SRDAMR_GRP1_EnableAutonomousClock(LL_SRDAMR_GRP1_PERIPH_LPTIM1AMEN);
LL_APB3_GRP1_ReleaseReset(LL_APB3_GRP1_PERIPH_LPTIM1);
```

ABESTM replied:
To be honest, I don't think the calls to `LL_APBx_GRPn_ReleaseReset()` do anything. This clears the reset bit in RCC and is usually called after setting this bit via `LL_APBx_GPRn_ForceReset()`. But I couldn't find any place that would set the reset bit and it is 0 by default. But perhaps I'm missing something?
The `LL_SRDAMR_GRP1_EnableAutonomousClock(LL_SRDAMR_GRP1_PERIPH_LPTIM1AMEN);` is still useful, since it allows LPTIM operation in autonomous mode.

ABESTM replied:
It seems the `LL_APBx_GRPn_ForceReset()` was present in initial commit of this file. But was later removed in commit 0856e9915500e2138772a5553ceabc2e6d78b7d0. However the `LL_APBx_GRPn_ReleaseReset()`.
I think it would be best to remove all `LL_APBx_GRPn_ReleaseReset()` calls and keep only `LL_SRDAMR_GRP1_EnableAutonomousClock(LL_SRDAMR_GRP1_PERIPH_LPTIM1AMEN);` for U5.
Should I do it as a separate PR, or separate commit inside PR is enough?

erwango replied:
A separate commit in this PR would be fine, thanks.

erwango commented on samples/boards/stm32/power_mgmt/serial_wakeup/boards/nucleo_h563zi.overlay:
```diff
@@ -0,0 +1,26 @@
+
+&usart3 {
+	/* Set domain clock to HSI to allow wakeup from Stop mode */
+    clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00040000>,
```
Please fix indentation

erwango reviewed (changes requested):
Thanks for this contribution. Looks fine but one request on the LPTIM driver changes

ABESTM committed:
drivers: timer: remove reset release for lptim

The LL_APBx_GRPn_ReleaseReset clears reset bit in RCC,
however this bit is reset by default. Code that set
this bit was removed in previous commit
0856e9915500e2138772a5553ceabc2e6d78b7d0
Keepint the autonomous clocks for STM32U5 family.

Signed-off-by: Adam Berlinger <adam.berlinger@st.com>

erwango reviewed (approved):
Great, thank you!
